pennylane/interfaces/batch/jax.py

```diff
@@ -0,0 +1,183 @@
+        gradient_kwargs=gradient_kwargs,
+        _n=_n,
+    )
+
+
+def _execute(
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+    jacs = None
+    total_size = np.sum([t.output_dim for t in tapes])
+    total_params = np.sum([len(p) for p in params])
+
+    @jax.custom_vjp
+    def wrapped_exec(params):
+        def wrapper(p):
+            new_tapes = []
+
+            for t, a in zip(tapes, p):
+                new_tapes.append(t.copy(copy_operations=True))
+                new_tapes[-1].set_parameters(a)
+
+            with qml.tape.Unwrap(*new_tapes):
+                res, jacs = execute_fn(new_tapes, **gradient_kwargs)
+
+            return np.stack(res)
+
+        res = host_callback.call(
+            wrapper, params, result_shape=jax.ShapeDtypeStruct((total_size, 1), jnp.float32)
+        )
+        return res
+
+    def wrapped_exec_fwd(params):
+        return wrapped_exec(params), params
+
+    def wrapped_exec_bwd(params, g):
+
+        # TODO: forward mode
+        if False:
+            return []
+        else:
+            module_name = getattr(inspect.getmodule(gradient_fn), "__name__", "")
+            if "pennylane.gradients" in module_name:
+
+                def non_diff_wrapper(args):
+                    """The derivative order is at the maximum. Compute the VJP
+                    in a non-differentiable manner to reduce overhead."""
+                    new_tapes = []
+                    p = args[:-1]
+                    dy = args[-1]
+
+                    for t, a in zip(tapes, p):
+                        new_tapes.append(t.copy(copy_operations=True))
+                        new_tapes[-1].set_parameters(a)
+                        new_tapes[-1].trainable_params = t.trainable_params
+
+                    vjp_tapes, processing_fn = qml.gradients.batch_vjp(
+                        new_tapes,
+                        dy,
+                        gradient_fn,
+                        reduction="append",
+                        gradient_kwargs=gradient_kwargs,
+                    )
+
+                    partial_res = execute_fn(vjp_tapes)[0]
+                    res = processing_fn(partial_res)
+                    return np.concatenate(res)
+
+                args = tuple(params) + (g,)
+                vjps = host_callback.call(
+                    non_diff_wrapper,
+                    args,
+                    result_shape=jax.ShapeDtypeStruct((total_params,), jnp.float32),
+                )
+
+                start = 0
+                res = []
+
+                for p in params:
+                    res.append(vjps[start : start + len(p)])
+                    start += len(p)
+                    if len(p) == 1:
+                        res[-1] = res[-1][0]
+
+                if res[0].ndim != 0:
+                    res = [[jnp.array(p) for p in res[0]]]

```

Comment:
What is happening in these two code blocks? Might be good to have some comments :)

Reply:
Added comments.